picard/util/__init__.py

```diff
@@ -891,3 +894,30 @@ def wildcards_to_regex_pattern(pattern):
         regex.append('\\[')
         regex.append(wildcards_to_regex_pattern(''.join(group[1:])))
     return ''.join(regex)
+
+
+def unique_numbered_title(default_title, existing_titles):
+    """Generate a new unique and numbered title
+       based on given default title and existing titles
+    """
+    escaped = re.escape(default_title)
+    regex = re.compile('^' + escaped + '(?: \\((\\d+)\\))?$')
+    count = 0
+    for title in existing_titles:
+        m = regex.match(title)
+        if m:
+            num = m.group(1)
+            if num is not None:
+                count = max(count, int(num))
+            else:
+                count += 1
+    return "{0} ({1})".format(default_title, count + 1)
+
+
+def get_base_title(title):
+    """Extract the base portion of a title,
+       removing the "(copy)" and number portion from the end.
+    """
+    re_text = "^(.*)\\s+{0}(\\s+\\(\\d*\\))?$".format(re.escape(_(DEFAULT_COPY_TEXT)))
```

Comment:
I would pass `suffix`:

```python
def get_base_title(title, suffix):
    escaped_suffix = re.escape(suffix)
    re_text = "^(.*)\\s+" + escaped_suffix + "(\\s+\\(\\d*\\))?$"
```

Then you can use it with:

```python
get_base_title(title, _(DEFAULT_COPY_TEXT))
```

And, in tests, with a constant string:

```python
get_base_title(title, 'copy')
```

Or even, `c?py` to ensure it is escaped.

Reply:
Great suggestion.  Thanks.  I did make one change though by making the `suffix` parameter optional and using the `DEFAULT_COPY_TEXT` value if it's missing.  Is that okay?  I know that it will break one of the tests if we ever change the value of `DEFAULT_COPY_TEXT`.  Should I remove that test, because (as you noted earlier) it's a bad idea to rely on translated texts in tests.

Reply:
>  Thanks. I did make one change though by making the `suffix` parameter optional and using the `DEFAULT_COPY_TEXT` value if it's missing. Is that okay?

Actually I think you should split this in two parts, a simple function for tests, and a convenient for code:

Near its uses:

```python
from picard.util import get_base_title_suffix

DEFAULT_COPY_TEXT = N_("(copy)")

def get_base_title(title):
   return get_base_title_suffix(title, _(DEFAULT_COPY_TEXT))
```

In `picard/util`:

```python
def get_base_title_suffix(title, suffix):
    escaped_suffix = re.escape(suffix)
...

In tests, use `get_base_title_suffix()`

Feel free to use better names ;)

Reply:
I see where that makes a lot of sense.  I'll make the changes.  Thanks for your patience.